src/cursor-doc/paredit.ts

```diff
@@ -792,6 +792,64 @@ export async function close(
   }
 }
 
+function onlyWhitespaceOnSameLineAsCursor(cursor: LispTokenCursor) {
+  cursor = cursor.clone();
+  while (!cursor.atStart()) {
+    switch (cursor.getPrevToken().type) {
+      case 'eol':
+        return true;
+      case 'ws':
+        cursor.previous();
+        continue;
+      default:
+        return false;
+    }
+  }
+}
+
+function backspaceOnWhitespaceEdit(doc: EditableDocument, cursor: LispTokenCursor) {
```

Comment:
My intuition here would be to check if all the text before the cursor on the same line is whitespace, and if so send a cursor `backwardWhitespace()` (without skipping comments), then replace the range we have moved with '' if the previous token is an open and ' ' otherwise. Maybe you considered that and there is a reason not to keep it that simple?

Reply:
Ok, I didn't know about `backwardWhitespace`, but one problem would be that it skips unlimited line endings. I can use it, but then I'd need to add a parameter to limit the number of 'eol' tokens to skip. Is that ok?

Here is an example from Cursive showing how we shouldn't kill all blank lines before the cursor:

https://user-images.githubusercontent.com/607986/170802133-e05c2403-efcf-4c66-bf71-bf4367fad8f2.mp4

Reply:
We try to preserve empty lines in Calva while typing as well, so that Cursive behaviour looks right to me.

Maybe what should be replaced is ”the text before the cursor on the same line + the newline before that”? The range would then be `prevPrevToken.startOffset` to `prevToken.endOffset`. (Haven't had my morning coffee yet, so might be missing something important here.) And as long as we're doing post-formatting we might get away with always replacing with a space.

Reply:
Thinking a bit more about this... if from prevPrevToken,startOffset, there is only whiteSpaceLeftOfCursor, and we do `getIndents()` from there, then we can replace the range with that many spaces and get rid of the post-formatting.